File: 백준/Silver/2343. 기타 레슨/기타 레슨.py
from sys import stdin, maxsize

input = stdin.readline

N, M = map(int, input().split())  # N: 강의의 수, M: 블루레이의 개수
lectures = list(map(int, input().split()))  # 강의 정보


def get_cnt(x):
    total = 0
    cnt = 0
    for i in range(N):
        # 현재 mid 값보다 크다면 -> 블루레이 하나 소모
        if total + lectures[i] > x:
            cnt += 1  # 블루레이 증가
            total = 0
        total += lectures[i]

    # 마지막 값이 남아있다면 -> 블루레이 하나 소모
    if total != 0:
        cnt += 1

    return cnt


def binary_search():
    left = max(lectures)  # 강의의 최대길이가 곧 최대 횟수를 만든다.
    right = sum(lectures)  # 모든 강의 시간의 합이 곧 최소 횟수를 만든다.
    ans = maxsize  # 정답 변수

    while left <= right:
        mid = (left + right) // 2
        # M개의 블루레이로 담을 수 없는 경우 -> mid 값을 늘려야한다.
        if M < get_cnt(mid):
            left = mid + 1

        # M개의 블루레이로 담을 수 있는 경우 -> mid 값을 낮춰야한다.
        else:
            right = mid - 1
            ans = min(ans, mid)

    return ans


print(binary_search())

File: Java/백준/Silver/2343. 기타 레슨/기타 레슨.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int N, M, MAX = 0, MIN = -1; // N: 강의의 개수, M: 녹화할 블루레이의 개수
    static int[] arr;

    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            MAX += arr[i];
            MIN = Math.max(MIN, arr[i]);
        }

        System.out.println(bs());
    }

    private static int bs() {

        int left = MIN;
        int right = MAX;
        int ans = Integer.MAX_VALUE;
        while (left <= right) {
            int sum = 0;
            int cnt = 0;
            int mid = (left + right) / 2;
            for (int i = 0; i < N; i++) {
                if (sum + arr[i] <= mid) {
                    if (i == N - 1)
                        cnt++;
                    sum += arr[i];
                } else {
                    cnt++;
                    sum = arr[i];
                    if (i == N - 1)
                        cnt++;
                }
            }
            // 목표 개수와 같거나 적게 나오는 경우 -> 블루레이 크기를 줄인다.
            if (cnt <= M) {
                ans = Math.min(ans, mid);
                right = mid - 1;
            // 목표 개수보다 더 많이 나오는 경우 -> 블루레이 크기를 늘린다.
            } else {
                left = mid + 1;
            }
        }

        return ans;
    }
}
